### DISCUSSION TITLE
Defi protocol - Lesson 14: Setup liquidations

### DISCUSSION THREAD

**Discussion by whxint** [Cyfrin/foundry-full-course-cu#1625] [Q&A]
**Liquidator pays someone's debt and burns its own DSC, but its state in the engine contract does not change.**

When the burn() function runs, the liquidated user's DSC amount decreases, but the liquidater's tokens are transferred to the contract and burned. Considering that the liquidater has a DSC of $100 dollars, it does not have any DSCs since they are all burned, but it continues to write $100 dollars in the s_DSCMinted variable of the contract. When liquidater comes later and mints $100, he will write $200 in the s_DSCMinted variable, but he will have $100 dollars.  It always has extra DSC. 

```
function liquidate(
        address collateral,
        address user,
        uint256 debtToCover
    )
        external
        moreThanZero(debtToCover)
        nonReentrant
    {
        uint256 startingUserHealthFactor = _healthFactor(user);
        if (startingUserHealthFactor >= MIN_HEALTH_FACTOR) {
            revert DSCEngine__HealthFactorOk();
        }
        // If covering 100 DSC, we need to $100 of collateral
        uint256 tokenAmountFromDebtCovered = getTokenAmountFromUsd(collateral, debtToCover);
        // And give them a 10% bonus
        // So we are giving the liquidator $110 of WETH for 100 DSC
        // We should implement a feature to liquidate in the event the protocol is insolvent
        // And sweep extra amounts into a treasury
        uint256 bonusCollateral = (tokenAmountFromDebtCovered * LIQUIDATION_BONUS) / LIQUIDATION_PRECISION;
        // Burn DSC equal to debtToCover
        // Figure out how much collateral to recover based on how much burnt
        _redeemCollateral(collateral, tokenAmountFromDebtCovered + bonusCollateral, user, msg.sender);
        _burnDsc(debtToCover, user, msg.sender);

        uint256 endingUserHealthFactor = _healthFactor(user);
        // This conditional should never hit, but just in case
        if (endingUserHealthFactor <= startingUserHealthFactor) {
            revert DSCEngine__HealthFactorNotImproved();
        }
        revertIfHealthFactorIsBroken(msg.sender);
    }
```

**Comment by EngrPips** [comment]
Hello @whxint, First, I want to clarify that a tutorial codebase sometimes does not attain a reality codebase standard. Knowing this would help us understand many things that are not considered in a tutorial codebase, as the goal is not to make a super correct codebase or bug-free codebase but to explain concepts and make people understand. 

Now, I think we should understand what you pointed out. What makes sense to me is that the protocol assumes that the liquidator had acquired `DSC` some other way apart from their debt with the protocol. So they are liquidating other users with `DSC` they acquired some different ways, and the `DSC` they get by minting from the protocol is still intact, and the accounting is still very much correct.

**Reply by whxint** [reply]
I don't quite understand what you mean, can you give me an example?

**Comment by EngrPips** [comment]
What I was pointing out is that the protocol assumes that anyone who tried to liquidate a user has more DSC than what they have minted from the protocol and that the extra DSC is enough to cover the liquidation; hence, that is why the DSC accounting of the liquidator is not affected by a liquidation. Honestly, this is just my thought about the current implementation. Maybe someone else might have a better understanding and make us understand better if I am wrong

**Reply by whxint** [reply]
Are you saying that it works this way so that the liquidator's liquidation is not affected ? But liquidater comes later and mints $100, he will write $200 in the s_DSCMinted variable, but he will have $100 dollars. It always has extra DSC. @PatrickAlphaC 

For example, the liquidator's state shows an extra $100 dollars, but he doesn't have that kind of money and he can't burn it. Someone else can send him $100 dollars and burn it. I don't know, it seemed ridiculous. Maybe it's really true.

**Reply by EngrPips** [reply]
I am saying that the liquidator mints $100 and has $100 s_DSCMinted on the protocol; if the liquidator liquidates a user, the DSC that would be used Is some DSC that was either sent to him by someone else or maybe he bought on an exchange and not what he minted on the engine. Hence why, his DSC balance is not reduced when he liquidates a user. I would like to let you know that @PatrickAlphaC  or the course didn't explicitly state this anywhere. I am just trying to make sense of why the DSC balance of the liquidator isn't accounted for in the DSCEngine after a successful liquidation call.

**Reply by whxint** [reply]
It is a DSC sent to him by someone else or perhaps purchased from the exchange. Therefore, when purging a user, their DSC balance will not decrease. You may be right in what you say. Liquidator can trigger the liquidity function by buying $100 directly from somewhere without making any deposit, but if someone is making mint transactions, the system will run into a bug. Extra amount of DSC is generated. It might have been overlooked because it wasn't a real project, or it might have worked that way. If the working mechanism is like this, it seems ridiculous.

**Reply by EngrPips** [reply]
The codebase is not perfect as it is not a production codebase, and you can open an issue on this Github repo to point out what you think needs improvement and how it should be done.

**Reply by whxint** [reply]
Well, even if it is not a bug, if Liquidator is actually someone who minted DSC, what do you think is the reason for such an incident? Why does he have extra DSC that he doesn't have? Because when Liquidator redeems, he will always have extra DSC. This point still seems strange to me, but there may be a logical explanation.

**Reply by EngrPips** [reply]
Honestly, we should not just conclude this is a design choice as it might be an error and needs to be corrected. I will say open an issue, clearly stating this issue you have found and what should be done to resolve it, and that would be a significant contribution to the course and an outstanding contribution from you as well.

**Comment by PatrickAlphaC** [comment]
Good question, two points.

The `s_DSCMinted` storage variable is actually more of a way to track how much _debt_ the protocol has. If I mint $100 in DSC, the protocol is keeping track of how much DSC it will eventually need to be repaid. 

To your point, let's look at a scenario.

1. `User A` puts down $200 of ETH in collateral and mints $100 of DSC. 
Status:
```
s_DSCMinted(User A) -> 100
s_collateralDeposited(User A)(ETH) -> 1 ETH ($200)
healthFactor(User A) -> Above 1
DSC in User A's wallet -> 100
```

2. The price of ETH plummets, and now the protocol has "bad debt." `User A` now has $125 of ETH as collateral and $100 of DSC minted. Since our collateralization ratio requires at least 2X as much ETH as DSC, we know the health factor of `User A` is below 1, and can be liquidated.
Status:
```
s_DSCMinted(User A) -> 100
s_collateralDeposited(User A)(ETH) -> 1 ETH ($125)
healthFactor(User A) -> Above 1
DSC in User A's wallet -> 100
```

3. `User B` comes along, and sees this, and wants to liquidate `User A`. They put down $250 of ETH of collateral, and mint $100 DSC to later pay off User A's debt

```
s_DSCMinted(User A) -> 100
s_collateralDeposited(User A)(ETH) -> 1 ETH ($125)
healthFactor(User A) -> Below 1
DSC in User A's wallet -> 100

s_DSCMinted(User A) -> 100
s_collateralDeposited(User A)(ETH) -> 2 ETH ($250)
healthFactor(User A) -> Above 1
DSC in User B's wallet -> 100
```

4. They now can liquidate User A. They will use their minted DSC to remove the "bad debt" from user A

```
s_DSCMinted(User A) -> 0 
s_collateralDeposited(User A)(ETH) -> 0
healthFactor(User A) -> Infinity (no debt)
DSC in User A's wallet -> 100

s_DSCMinted(User A) -> 100 (This is the debt they previously had)
s_collateralDeposited(User A)(ETH) -> 2 ETH + 1 ETH Liquidated
healthFactor(User A) -> Above 1
DSC in User B's wallet -> 0 (100 used to pay off User A's)
```


## Also

This is just a dummy stablecoin for learning purposes. It's not meant to be exhaustive. You can see a formal competitive audit the codebase went under [here and see a lot of the issues it has.](https://www.codehawks.com/contests/cljx3b9390009liqwuedkn0m0)

Let me know if this solves your issue.

**Reply by whxint** [reply]
Part 4 is wrong again according to your code.

Is this your audited contract ? ->https://github.com/Cyfrin/2023-07-foundry-defi-stablecoin

**Reply by EngrPips** [reply]
That is why we stated that the codebase isn't a production codebase but a tutorial codebase. It is expected to have flaws as nobody would take all the time in the world to think while trying to build a codebase to teach people stuff, and it would never be deployed. So, it is always a good chance for the student to find pieces that need to be done correctly or are not being considered, then report the issues and come up with a solution to the problem.

**Reply by whxint** [reply]
I haven't reviewed the audited code, but I will. I guess there shouldn't be any errors in the audited code because the audited code is the final version of the project. Nobody wants to give away 15 thousand dollars for nothing.

**Reply by EngrPips** [reply]
The audited code was more of helping kickstart many of Patrick's student auditing careers and not that he highly intended to push the code to production. That's my conclusion, not that he stated it explicitly anywhere.
